File: packages/js-sdk/tests/deliverable.test.ts
/**
 * Deliverable Module Tests
 *
 * Tests for SDK operations:
 * - listDeliverables
 * - generateDeliverable
 * - getDeliverableDetails
 * - updateDeliverableInfo
 * - deleteDeliverable
 * - downloadSourceFile
 * - downloadPDF
 */

import { Deliverable } from "../src/modules/deliverable";
import { HttpClient } from "../src/http";

// Mock the HttpClient
jest.mock("../src/http");

const MockedHttpClient = HttpClient as jest.MockedClass<typeof HttpClient>;

describe("Deliverable Module", () => {
  beforeEach(() => {
    jest.clearAllMocks();
    // Reset static client
    (Deliverable as any).client = undefined;
  });

  describe("configure", () => {
    it("should configure the client with API key and org ID", () => {
      Deliverable.configure({
        apiKey: "test-api-key",
        orgId: "test-org-id",
      });
      expect(MockedHttpClient).toHaveBeenCalledWith({
        apiKey: "test-api-key",
        accessToken: undefined,
        orgId: "test-org-id",
        baseUrl: undefined,
        skipSenderValidation: true,
      });
    });

    it("should configure with custom base URL", () => {
      Deliverable.configure({
        apiKey: "test-api-key",
        orgId: "test-org-id",
        baseUrl: "https://custom-api.example.com",
      });
      expect(MockedHttpClient).toHaveBeenCalledWith({
        apiKey: "test-api-key",
        accessToken: undefined,
        orgId: "test-org-id",
        baseUrl: "https://custom-api.example.com",
        skipSenderValidation: true,
      });
    });

    it("should configure with access token instead of API key", () => {
      Deliverable.configure({
        accessToken: "oauth-token",
        orgId: "test-org-id",
      });
      expect(MockedHttpClient).toHaveBeenCalledWith({
        apiKey: undefined,
        accessToken: "oauth-token",
        orgId: "test-org-id",
        baseUrl: undefined,
        skipSenderValidation: true,
      });
    });
  });

  describe("listDeliverables", () => {
    it("should list deliverables with default options", async () => {
      const mockResponse = {
        results: [
          {
            id: "del-1",
            name: "Contract A",
            description: "A contract",
            templateId: "tmpl-1",
            createdBy: "user-1",
            isActive: true,
            createdOn: "2024-01-15T14:12:10.721Z",
            updatedOn: "2024-01-15T14:13:45.724Z",
          },
        ],
        totalRecords: 1,
      };

      MockedHttpClient.prototype.get = jest.fn().mockResolvedValue(mockResponse);
      Deliverable.configure({ apiKey: "test-key", orgId: "org-1" });

      const result = await Deliverable.listDeliverables();

      expect(result.results).toHaveLength(1);
      expect(result.totalRecords).toBe(1);
      expect(MockedHttpClient.prototype.get).toHaveBeenCalledWith(
        "/v1/deliverable",
        {}
      );
    });

    it("should pass query parameters", async () => {
      const mockResponse = { results: [], totalRecords: 0 };
      MockedHttpClient.prototype.get = jest.fn().mockResolvedValue(mockResponse);
      Deliverable.configure({ apiKey: "test-key", orgId: "org-1" });

      await Deliverable.listDeliverables({
        limit: 10,
        offset: 20,
        query: "contract",
        showTags: true,
      });

      expect(MockedHttpClient.prototype.get).toHaveBeenCalledWith(
        "/v1/deliverable",
        {
          limit: 10,
          offset: 20,
          query: "contract",
          showTags: true,
        }
      );
    });
  });

  describe("generateDeliverable", () => {
    it("should generate a deliverable from a template", async () => {
      const mockResponse = {
        results: {
          deliverable: {
            id: "del-new",
            name: "Employee Contract - John Smith",
            description: "Employment contract",
            templateId: "tmpl-1",
            createdBy: "user-1",
            createdOn: "2024-01-15T14:12:10.721Z",
            updatedOn: "2024-01-15T14:12:10.721Z",
            isActive: true,
            defaultFont: "",
            fonts: null,
          },
        },
      };

      MockedHttpClient.prototype.post = jest.fn().mockResolvedValue(mockResponse);
      Deliverable.configure({ apiKey: "test-key", orgId: "org-1" });

      const result = await Deliverable.generateDeliverable({
        name: "Employee Contract - John Smith",
        templateId: "tmpl-1",
        variables: [
          { placeholder: "{EmployeeName}", text: "John Smith", mimeType: "text" },
          { placeholder: "{CompanyName}", text: "TechCorp Inc.", mimeType: "text" },
        ],
        tags: ["hr", "contract"],
      });

      expect(result.results.deliverable.id).toBe("del-new");
      expect(MockedHttpClient.prototype.post).toHaveBeenCalledWith(
        "/v1/deliverable",
        {
          name: "Employee Contract - John Smith",
          templateId: "tmpl-1",
          variables: [
            { placeholder: "{EmployeeName}", text: "John Smith", mimeType: "text" },
            { placeholder: "{CompanyName}", text: "TechCorp Inc.", mimeType: "text" },
          ],
          tags: ["hr", "contract"],
        }
      );
    });

    it("should support variable stacks for repeating content", async () => {
      const mockResponse = {
        results: {
          deliverable: {
            id: "del-stack",
            name: "Project Report",
            description: "",
            templateId: "tmpl-2",
            createdBy: "user-1",
            isActive: true,
            createdOn: "2024-01-15T14:12:10.721Z",
            updatedOn: "2024-01-15T14:12:10.721Z",
          },
        },
      };

      MockedHttpClient.prototype.post = jest.fn().mockResolvedValue(mockResponse);
      Deliverable.configure({ apiKey: "test-key", orgId: "org-1" });

      const result = await Deliverable.generateDeliverable({
        name: "Project Report",
        templateId: "tmpl-2",
        variables: [
          {
            placeholder: "{ProjectPhase}",
            mimeType: "html",
            variableStack: {
              "0": { text: "<p>Phase 1: Assess</p>", mimeType: "html" },
              "1": { text: "<p>Phase 2: Remediate</p>", mimeType: "html" },
            },
          },
        ],
      });

      expect(result.results.deliverable.id).toBe("del-stack");
    });
  });

  describe("getDeliverableDetails", () => {
    it("should get full deliverable details", async () => {
      const mockResponse = {
        results: {
          id: "del-1",
          name: "Employee Contract",
          description: "Employment contract",
          templateId: "tmpl-1",
          templateName: "Contract Template",
          templateNotDeleted: true,
          defaultFont: "Arial",
          fonts: [{ name: "Arial", usage: "body" }],
          email: "admin@company.com",
          fileSize: 287456,
          fileType: "application/vnd.openxmlformats-officedocument.wordprocessingml.document",
          isActive: true,
          createdBy: "user-1",
          createdOn: "2024-01-15T14:12:10.721Z",
          updatedOn: "2024-01-15T14:13:45.724Z",
          variables: [
            { placeholder: "{EmployeeName}", text: "John Smith", mimeType: "text" },
          ],
        },
      };

      MockedHttpClient.prototype.get = jest.fn().mockResolvedValue(mockResponse);
      Deliverable.configure({ apiKey: "test-key", orgId: "org-1" });

      const result = await Deliverable.getDeliverableDetails("del-1");

      expect(result.id).toBe("del-1");
      expect(result.templateName).toBe("Contract Template");
      expect(result.variables).toHaveLength(1);
      expect(MockedHttpClient.prototype.get).toHaveBeenCalledWith(
        "/v1/deliverable/del-1",
        {}
      );
    });

    it("should pass showTags option", async () => {
      const mockResponse = {
        results: {
          id: "del-1",
          name: "Contract",
          tags: [{ id: "tag-1", name: "hr" }],
          isActive: true,
          createdBy: "user-1",
          createdOn: "2024-01-15T14:12:10.721Z",
          updatedOn: "2024-01-15T14:13:45.724Z",
        },
      };

      MockedHttpClient.prototype.get = jest.fn().mockResolvedValue(mockResponse);
      Deliverable.configure({ apiKey: "test-key", orgId: "org-1" });

      const result = await Deliverable.getDeliverableDetails("del-1", { showTags: true });

      expect(result.tags).toHaveLength(1);
      expect(MockedHttpClient.prototype.get).toHaveBeenCalledWith(
        "/v1/deliverable/del-1",
        { showTags: true }
      );
    });
  });

  describe("updateDeliverableInfo", () => {
    it("should update deliverable name and tags", async () => {
      const mockResponse = {
        message: "Deliverable updated successfully",
        deliverableId: "del-1",
      };

      MockedHttpClient.prototype.patch = jest.fn().mockResolvedValue(mockResponse);
      Deliverable.configure({ apiKey: "test-key", orgId: "org-1" });

      const result = await Deliverable.updateDeliverableInfo("del-1", {
        name: "Updated Contract",
        tags: ["hr", "finalized"],
      });

      expect(result.message).toBe("Deliverable updated successfully");
      expect(result.deliverableId).toBe("del-1");
      expect(MockedHttpClient.prototype.patch).toHaveBeenCalledWith(
        "/v1/deliverable/del-1",
        { name: "Updated Contract", tags: ["hr", "finalized"] }
      );
    });
  });

  describe("deleteDeliverable", () => {
    it("should soft-delete a deliverable", async () => {
      const mockResponse = {
        message: "Deliverable deleted successfully",
        deliverableId: "del-1",
      };

      MockedHttpClient.prototype.delete = jest.fn().mockResolvedValue(mockResponse);
      Deliverable.configure({ apiKey: "test-key", orgId: "org-1" });

      const result = await Deliverable.deleteDeliverable("del-1");

      expect(result.message).toBe("Deliverable deleted successfully");
      expect(MockedHttpClient.prototype.delete).toHaveBeenCalledWith(
        "/v1/deliverable/del-1"
      );
    });
  });

  describe("downloadSourceFile", () => {
    it("should download source file as ArrayBuffer", async () => {
      const mockArrayBuffer = new ArrayBuffer(1024);

      MockedHttpClient.prototype.getRaw = jest.fn().mockResolvedValue(mockArrayBuffer);
      Deliverable.configure({ apiKey: "test-key", orgId: "org-1" });

      const result = await Deliverable.downloadSourceFile("del-1");

      expect(result).toBe(mockArrayBuffer);
      expect(MockedHttpClient.prototype.getRaw).toHaveBeenCalledWith(
        "/v1/deliverable/file/del-1"
      );
    });
  });

  describe("downloadPDF", () => {
    it("should download PDF as ArrayBuffer", async () => {
      const mockArrayBuffer = new ArrayBuffer(2048);

      MockedHttpClient.prototype.getRaw = jest.fn().mockResolvedValue(mockArrayBuffer);
      Deliverable.configure({ apiKey: "test-key", orgId: "org-1" });

      const result = await Deliverable.downloadPDF("del-1");

      expect(result).toBe(mockArrayBuffer);
      expect(MockedHttpClient.prototype.getRaw).toHaveBeenCalledWith(
        "/v1/deliverable/file/pdf/del-1"
      );
    });
  });

  describe("Error Handling", () => {
    it("should throw error when not configured", async () => {
      await expect(Deliverable.listDeliverables()).rejects.toThrow(
        "Deliverable not configured"
      );
    });

    it("should handle API errors gracefully", async () => {
      const apiError = {
        statusCode: 404,
        message: "Deliverable not found",
      };

      MockedHttpClient.prototype.get = jest.fn().mockRejectedValue(apiError);
      Deliverable.configure({ apiKey: "test-key", orgId: "org-1" });

      await expect(
        Deliverable.getDeliverableDetails("invalid-id")
      ).rejects.toEqual(apiError);
    });

    it("should handle validation errors", async () => {
      const validationError = {
        statusCode: 400,
        message: '"name" length must be at least 3 characters long',
      };

      MockedHttpClient.prototype.post = jest.fn().mockRejectedValue(validationError);
      Deliverable.configure({ apiKey: "test-key", orgId: "org-1" });

      await expect(
        Deliverable.generateDeliverable({
          name: "AB",
          templateId: "tmpl-1",
          variables: [],
        })
      ).rejects.toEqual(validationError);
    });
  });
});

File: packages/go-sdk/deliverable_test.go
package turbodocx

import (
	"context"
	"encoding/json"
	"net/http"
	"net/http/httptest"
	"testing"

	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
)

func newDeliverableTestClient(t *testing.T, handler http.HandlerFunc) *DeliverableClient {
	server := httptest.NewServer(handler)
	t.Cleanup(server.Close)

	client, err := NewDeliverableClientOnly(ClientConfig{
		APIKey:  "test-api-key",
		OrgID:   "test-org-id",
		BaseURL: server.URL,
	})
	require.NoError(t, err)
	return client
}

func TestNewDeliverableClientOnly(t *testing.T) {
	t.Run("creates client without sender email", func(t *testing.T) {
		client, err := NewDeliverableClientOnly(ClientConfig{
			APIKey:  "test-key",
			OrgID:   "test-org",
			BaseURL: "https://api.example.com",
		})
		require.NoError(t, err)
		assert.NotNil(t, client)
	})

	t.Run("requires API key", func(t *testing.T) {
		_, err := NewDeliverableClientOnly(ClientConfig{
			OrgID: "test-org",
		})
		require.Error(t, err)
	})

	t.Run("requires org ID", func(t *testing.T) {
		_, err := NewDeliverableClientOnly(ClientConfig{
			APIKey: "test-key",
		})
		require.Error(t, err)
		_, ok := err.(*AuthenticationError)
		assert.True(t, ok)
	})
}

func TestDeliverableClient_ListDeliverables(t *testing.T) {
	t.Run("lists deliverables", func(t *testing.T) {
		client := newDeliverableTestClient(t, func(w http.ResponseWriter, r *http.Request) {
			assert.Equal(t, "GET", r.Method)
			assert.Contains(t, r.URL.Path, "/v1/deliverable")
			assert.Equal(t, "5", r.URL.Query().Get("limit"))
			assert.Equal(t, "true", r.URL.Query().Get("showTags"))

			w.Header().Set("Content-Type", "application/json")
			json.NewEncoder(w).Encode(DeliverableListResponse{
				Results:      []DeliverableRecord{{ID: "d1", Name: "Contract A"}},
				TotalRecords: 1,
			})
		})

		result, err := client.ListDeliverables(context.Background(), &ListDeliverablesOptions{
			Limit:    5,
			ShowTags: true,
		})
		require.NoError(t, err)
		assert.Equal(t, 1, result.TotalRecords)
		assert.Equal(t, "d1", result.Results[0].ID)
	})

	t.Run("lists without options", func(t *testing.T) {
		client := newDeliverableTestClient(t, func(w http.ResponseWriter, r *http.Request) {
			assert.Equal(t, "/v1/deliverable", r.URL.Path)
			assert.Empty(t, r.URL.RawQuery)

			w.Header().Set("Content-Type", "application/json")
			json.NewEncoder(w).Encode(DeliverableListResponse{
				Results:      []DeliverableRecord{},
				TotalRecords: 0,
			})
		})

		result, err := client.ListDeliverables(context.Background(), nil)
		require.NoError(t, err)
		assert.Equal(t, 0, result.TotalRecords)
	})
}

func TestDeliverableClient_GenerateDeliverable(t *testing.T) {
	t.Run("generates deliverable", func(t *testing.T) {
		client := newDeliverableTestClient(t, func(w http.ResponseWriter, r *http.Request) {
			assert.Equal(t, "POST", r.Method)
			assert.Equal(t, "/v1/deliverable", r.URL.Path)
			assert.Equal(t, "Bearer test-api-key", r.Header.Get("Authorization"))

			var body CreateDeliverableRequest
			json.NewDecoder(r.Body).Decode(&body)
			assert.Equal(t, "Employee Contract", body.Name)
			assert.Equal(t, "tmpl-1", body.TemplateID)
			assert.Len(t, body.Variables, 1)

			w.Header().Set("Content-Type", "application/json")
			json.NewEncoder(w).Encode(CreateDeliverableResponse{
				Results: struct {
					Deliverable DeliverableRecord `json:"deliverable"`
				}{
					Deliverable: DeliverableRecord{ID: "new-d-id", Name: "Employee Contract"},
				},
			})
		})

		result, err := client.GenerateDeliverable(context.Background(), &CreateDeliverableRequest{
			Name:       "Employee Contract",
			TemplateID: "tmpl-1",
			Variables: []DeliverableVariable{
				{Placeholder: "{Name}", Text: "John", MimeType: "text"},
			},
			Tags:         []string{"hr"},
		})
		require.NoError(t, err)
		assert.Equal(t, "new-d-id", result.Results.Deliverable.ID)
	})
}

func TestDeliverableClient_GetDeliverableDetails(t *testing.T) {
	t.Run("gets deliverable details", func(t *testing.T) {
		client := newDeliverableTestClient(t, func(w http.ResponseWriter, r *http.Request) {
			assert.Equal(t, "GET", r.Method)
			assert.Equal(t, "/v1/deliverable/d1", r.URL.Path)
			assert.Equal(t, "true", r.URL.Query().Get("showTags"))

			w.Header().Set("Content-Type", "application/json")
			json.NewEncoder(w).Encode(GetDeliverableResponse{
				Results: DeliverableRecord{
					ID:           "d1",
					Name:         "Contract A",
					TemplateName: "Standard",
					Tags:         []Tag{{ID: "t1", Label: "hr"}},
				},
			})
		})

		result, err := client.GetDeliverableDetails(context.Background(), "d1", &GetDeliverableOptions{ShowTags: true})
		require.NoError(t, err)
		assert.Equal(t, "d1", result.ID)
		assert.Equal(t, "Contract A", result.Name)
		assert.Len(t, result.Tags, 1)
	})
}

func TestDeliverableClient_UpdateDeliverableInfo(t *testing.T) {
	t.Run("updates deliverable", func(t *testing.T) {
		client := newDeliverableTestClient(t, func(w http.ResponseWriter, r *http.Request) {
			assert.Equal(t, "PATCH", r.Method)
			assert.Equal(t, "/v1/deliverable/d1", r.URL.Path)

			w.Header().Set("Content-Type", "application/json")
			json.NewEncoder(w).Encode(UpdateDeliverableResponse{
				Message:       "Deliverable updated successfully",
				DeliverableID: "d1",
			})
		})

		tags := []string{"finalized"}
		result, err := client.UpdateDeliverableInfo(context.Background(), "d1", &UpdateDeliverableRequest{
			Name: "Updated Name",
			Tags: &tags,
		})
		require.NoError(t, err)
		assert.Equal(t, "Deliverable updated successfully", result.Message)
	})
}

func TestDeliverableClient_DeleteDeliverable(t *testing.T) {
	t.Run("deletes deliverable", func(t *testing.T) {
		client := newDeliverableTestClient(t, func(w http.ResponseWriter, r *http.Request) {
			assert.Equal(t, "DELETE", r.Method)
			assert.Equal(t, "/v1/deliverable/d1", r.URL.Path)

			w.Header().Set("Content-Type", "application/json")
			json.NewEncoder(w).Encode(DeleteDeliverableResponse{
				Message:       "Deliverable deleted successfully",
				DeliverableID: "d1",
			})
		})

		result, err := client.DeleteDeliverable(context.Background(), "d1")
		require.NoError(t, err)
		assert.Equal(t, "Deliverable deleted successfully", result.Message)
	})
}

func TestDeliverableClient_DownloadSourceFile(t *testing.T) {
	t.Run("downloads source file", func(t *testing.T) {
		mockContent := []byte("PK\x03\x04mock-docx-content")
		client := newDeliverableTestClient(t, func(w http.ResponseWriter, r *http.Request) {
			assert.Equal(t, "GET", r.Method)
			assert.Equal(t, "/v1/deliverable/file/d1", r.URL.Path)
			w.Write(mockContent)
		})

		result, err := client.DownloadSourceFile(context.Background(), "d1")
		require.NoError(t, err)
		assert.Equal(t, mockContent, result)
	})
}

func TestDeliverableClient_DownloadPDF(t *testing.T) {
	t.Run("downloads PDF", func(t *testing.T) {
		mockContent := []byte("%PDF-mock-content")
		client := newDeliverableTestClient(t, func(w http.ResponseWriter, r *http.Request) {
			assert.Equal(t, "GET", r.Method)
			assert.Equal(t, "/v1/deliverable/file/pdf/d1", r.URL.Path)
			w.Write(mockContent)
		})

		result, err := client.DownloadPDF(context.Background(), "d1")
		require.NoError(t, err)
		assert.Equal(t, mockContent, result)
	})
}

